File: backend/app/schemas/flashcard.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class FlashcardCreate(BaseModel):
    front: str = Field(min_length=1, max_length=1000)
    back: str = Field(min_length=1, max_length=1000)


class FlashcardUpdate(BaseModel):
    front: str | None = Field(default=None, min_length=1, max_length=1000)
    back: str | None = Field(default=None, min_length=1, max_length=1000)


class FlashcardGenerateRequest(BaseModel):
    count: int = Field(default=10, ge=1, le=20)


class FlashcardReviewRequest(BaseModel):
    result: str = Field(pattern="^(again|good)$")


class FlashcardProgressRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    status: str
    times_reviewed: int
    times_correct: int
    correct_streak: int
    last_reviewed_at: datetime | None


class FlashcardRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    deck_id: uuid.UUID
    front: str
    back: str
    created_at: datetime
    updated_at: datetime
    progress: FlashcardProgressRead


class DeckCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    subject_id: uuid.UUID | None = None
    description: str | None = Field(default=None, max_length=2000)


class DeckUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    subject_id: uuid.UUID | None = None
    description: str | None = Field(default=None, max_length=2000)


class DeckRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    subject_id: uuid.UUID | None
    title: str
    description: str | None
    card_count: int
    mastered_count: int
    mastery_percentage: int
    created_at: datetime
    updated_at: datetime


class DeckDetail(DeckRead):
    flashcards: list[FlashcardRead]

File: frontend/src/types/flashcard.ts
export type MasteryStatus = 'new' | 'learning' | 'mastered'
export type ReviewResult = 'again' | 'good'

export interface FlashcardProgress {
  status: MasteryStatus
  times_reviewed: number
  times_correct: number
  correct_streak: number
  last_reviewed_at: string | null
}

export interface Flashcard {
  id: string
  deck_id: string
  front: string
  back: string
  created_at: string
  updated_at: string
  progress: FlashcardProgress
}

export interface Deck {
  id: string
  subject_id: string | null
  title: string
  description: string | null
  card_count: number
  mastered_count: number
  mastery_percentage: number
  created_at: string
  updated_at: string
}

export interface DeckDetail extends Deck {
  flashcards: Flashcard[]
}

export interface DeckCreateRequest {
  title: string
  subject_id?: string | null
  description?: string | null
}

export interface DeckUpdateRequest {
  title?: string
  subject_id?: string | null
  description?: string | null
}

export interface FlashcardCreateRequest {
  front: string
  back: string
}

export interface FlashcardUpdateRequest {
  front?: string
  back?: string
}

export interface FlashcardGenerateRequest {
  count?: number
}

export interface FlashcardReviewRequest {
  result: ReviewResult
}